src/libsystemd/sd-event/event-util.c

```diff
@@ -177,3 +180,88 @@ dual_timestamp* event_dual_timestamp_now(sd_event *e, dual_timestamp *ts) {
         assert_se(sd_event_now(e, CLOCK_MONOTONIC, &ts->monotonic) >= 0);
         return ts;
 }
+
+void event_source_unref_many(sd_event_source **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
+                sd_event_source_unref(*v);
+
+        free(array);
+}
+
+static int event_forward_signal_callback(sd_event_source *s, const struct signalfd_siginfo *ssi, void *userdata) {
+        sd_event_source *child = ASSERT_PTR(userdata);
+
+        assert(ssi);
+
+        siginfo_t si = {
+                .si_signo = ssi->ssi_signo,
+                /* We include some extra information to indicate the signal was forwarded and originally a positive
+                 * value since we can only set negative values ourselves as positive values are prohibited by the
+                 * kernel. */
+                .si_code = (ssi->ssi_code == 0 || ssi->ssi_code & (SI_FLAG_FORWARD|SI_FLAG_POSITIVE)) ? INT_MIN :
```

Comment:
why map SI_USER to IN_MIN? the whole point of adding SI_FLAG_FORWARD was that we can map SI_USER to something recognizable, i.e. SI_USER|SI_FLAG_FORWARD, but here you map it to IN_MIN anyway for no reason?

Reply:
@poettering 0 + flags the flags is a positive value, the value has to be negative otherwise it's blocked by the kernel

Reply:
dang, right. maybe shift do "-ssi_code-1" rather than "-ssi_code" then? kinda sucks, but it means we can reverse the operation nicely.